exec/cnex/lib/mmap_posix.c

```diff
@@ -0,0 +1,184 @@
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "cell.h"
+#include "exec.h"
+#include "nstring.h"
+#include "number.h"
+#include "object.h"
+#include "stack.h"
+
+
+typedef struct tagTMmapObject {
+    int fd;
+    size_t len;
+    char *view;
+} TMmapObject;
+
+
+static TMmapObject *check_file(TExecutor *exec, Object *pp)
+{
+    if (pp == NULL || pp->ptr == NULL || ((TMmapObject*)(pp->ptr))->fd < 0) {
+        exec->rtl_raise(exec, "MmapException.InvalidFile", "", BID_ZERO);
+        return NULL;
+    }
+    return pp->ptr;
+}
+
+void object_releaseMMapObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                free(o->ptr);
+            }
+        }
+    }
+}
+
+Cell *object_mmapObjectToString(Object *self)
+{
+    char s[32];
+    snprintf(s, 32, "<mmap %" PRId64 ">", (int64_t)((TMmapObject*)self->ptr)->fd);
+    Cell *r = cell_fromCString(s);
+    return r;
+}
+
+Object *object_createMMapObject(TMmapObject *p)
+{
+    Object *r = object_createObject();
+    r->ptr = p;
+    r->release = object_releaseMMapObject;
+    r->toString = object_mmapObjectToString;
+    return r;
+}
+
+
+
+
+void mmap_close(TExecutor *exec)
+{
+    Object *pf = top(exec->stack)->object;
+
+    TMmapObject *f = check_file(exec, pf);
+    if (f == NULL) {
+        pop(exec->stack);
+        return;
+    }
+
+    munmap(f->view, f->len);
+    close(f->fd);
+    f->fd = -1;
+
+    pop(exec->stack);
+}
+
+void mmap_open(TExecutor *exec)
+{
+    pop(exec->stack);
+    char *name = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    char err[256];
+    TMmapObject *f = malloc(sizeof(TMmapObject));
+
+    f->fd = open(name, O_RDONLY);
+    if (f->fd < 0) {
+        int e = errno;
+        free(f);
+        free(name);
+        snprintf(err, 256, "open: error (%d) %s", e, strerror(e));
+        exec->rtl_raise(exec, "OpenFileException", err, BID_ZERO);
+        return;
+    }
+
+    struct stat st;
+    fstat(f->fd, &st);
+    f->len = st.st_size;
+    f->view = (char *)mmap(NULL, f->len, PROT_READ, MAP_PRIVATE, f->fd, 0);
+    if (f->view == MAP_FAILED) {
+        int e = errno;
+        close(f->fd);
+        free(f);
+        free(name);
+        snprintf(err, 256, "mmap: error (%d) %s", e, strerror(e));
+        exec->rtl_raise(exec, "OpenFileException", err, BID_ZERO);
+        return;
+    }
+
+    push(exec->stack, cell_fromObject(object_createMMapObject(f)));
+}
+
+void mmap_read(TExecutor *exec)
+{
+    Number count = top(exec->stack)->number; pop(exec->stack);
+    Number offset = top(exec->stack)->number; pop(exec->stack);
+    Object *pf = top(exec->stack)->object;  pop(exec->stack);
+    Cell *ret = cell_createStringCell(0);
+    ret->type = cBytes;
+
+    TMmapObject *f = check_file(exec, pf);
+
+    uint64_t o = number_to_uint64(offset);
+    if (o >= f->len) {
+        ret->type = cBytes;
```

Comment:
Seems like a redundant assignment, since you've already done that a few lines above.

Reply:
Removed redundant assignment.